src/esp/io/URDFParser.cpp

```diff
@@ -0,0 +1,1058 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// Code adapted from Bullet3/examples/Importers/ImportURDFDemo ...
+
+#include <Corrade/Utility/DebugStl.h>
+#include <Corrade/Utility/Directory.h>
+#include <Corrade/Utility/String.h>
+#include <Magnum/Math/Quaternion.h>
+#include <iostream>
+
+#include "URDFParser.h"
+
+#include "tinyxml2/tinyxml2.h"
+
+// using namespace tinyxml2;
+using tinyxml2::XMLDocument;
+using tinyxml2::XMLElement;
+
+namespace Cr = Corrade;
+namespace Mn = Magnum;
+
+namespace esp {
+namespace io {
+namespace URDF {
+
+bool Parser::parseURDF(std::shared_ptr<Model>& urdfModel,
+                       const std::string& filename) {
+  Mn::Debug silence{logMessages ? &std::cout : nullptr};
+
+  // override the previous model with a fresh one
+  urdfModel = std::make_shared<Model>();
+  sourceFilePath_ = filename;
+  urdfModel->m_sourceFile = filename;
+
+  std::string xmlString = Corrade::Utility::Directory::readString(filename);
+
+  XMLDocument xml_doc;
+  xml_doc.Parse(xmlString.c_str());
+  if (xml_doc.Error()) {
+    Mn::Debug{}
+        << "Parser::parseURDF - XML parse error, aborting URDF parse/load.";
+    return false;
+  }
+  Mn::Debug{} << "Parser::parseURDF - XML parsed starting URDF parse/load.";
+
+  XMLElement* robot_xml = xml_doc.FirstChildElement("robot");
+  if (!robot_xml) {
+    Mn::Debug{} << "E - expected a robot element";
+    return false;
+  }
+
+  // Get robot name
+  const char* name = robot_xml->Attribute("name");
+  if (!name) {
+    Mn::Debug{} << "E - expected a name for robot";
+    return false;
+  }
+  urdfModel->m_name = name;
+
+  // Get all Material elements
+  for (XMLElement* material_xml = robot_xml->FirstChildElement("material");
+       material_xml;
+       material_xml = material_xml->NextSiblingElement("material")) {
+    Material material;
+
+    parseMaterial(material, material_xml);
+
+    if (urdfModel->m_materials.count(material.m_name) == 0) {
+      urdfModel->m_materials[material.m_name] =
+          std::make_shared<Material>(material);
+    } else {
+      Mn::Debug{} << "W - Duplicate material";
+    }
+  }
+
+  // Get all link elements including shapes
+  for (XMLElement* link_xml = robot_xml->FirstChildElement("link"); link_xml;
+       link_xml = link_xml->NextSiblingElement("link")) {
+    std::shared_ptr<Link> link = std::make_shared<Link>();
+
+    if (parseLink(urdfModel, *link, link_xml)) {
+      if (urdfModel->m_links.count(link->m_name) != 0u) {
+        Mn::Debug{} << "E - Link name is not unique, link names "
```

Comment:
Again, these error messages could specify the relevant file and also print to an error stream, not Mn::Debug.